---
title: "Beta-binomial test for NGS data"
author: "Yassine Souilmi"
date: "December 10, 2015"
output: ioslides_presentation
runtime: shiny
---

## Exercise
- NGS (RNA-Seq) genrates count data for quantitative analysis
- Assuming a negative binomial distribution
- Input data is in tsv (tab delimited format)
- The total the total count for gene (t) is: 
    
    $$gene(t) = isoform(m) + isoform(t)$$
    where: $t$ is normal and $m$ is malignant

## Load the data

We start by loading the data from the tab delimited file:
```{r, include=FALSE}
library(dplyr)  # for data manipulation
library(ggvis)  # for visualizaiton 
```
```{r}
# Read the raw_data.tsv file
raw_data <- read.table("raw_data.tsv", header = T, fill = T)
# Make sure the data was imported in a dplyr compatible data frame
raw <- raw_data %>% as_data_frame()
```

## Check the dataframe formating
```{r}
raw %>% head()
```
## â€¦ and content

```{r}
raw %>% sapply(class) # get the class of each column
```

## 

## Shiny Presentation

This R Markdown presentation is made interactive using Shiny. The viewers of the presentation can change the assumptions underlying what's presented and see the results immediately. 

To learn more, see [Interactive Documents](http://rmarkdown.rstudio.com/authoring_shiny.html).

## Slide with Interactive Plot

```{r, echo=FALSE}
inputPanel(
  selectInput("n_breaks", label = "Number of bins:",
              choices = c(10, 20, 35, 50), selected = 20),
  
  sliderInput("bw_adjust", label = "Bandwidth adjustment:",
              min = 0.2, max = 2, value = 1, step = 0.2)
)

renderPlot({
  hist(faithful$eruptions, probability = TRUE, breaks = as.numeric(input$n_breaks),
       xlab = "Duration (minutes)", main = "Geyser eruption duration")
  
  dens <- density(faithful$eruptions, adjust = input$bw_adjust)
  lines(dens, col = "blue")
})
```



